### PR TITLE
WIP: Helm chart

### PR DESCRIPTION
**Pull request by shaunc**
## Summary

This helm chart is a work in progress (not for merging -- except into separate branch if appropriate). Will create database and init with schema. All services come up except Apollo.

**If you have some time, please give me some help with apollo.** Currently failing with:
```
kc logs --follow prefect-server-apollo-bb688cf9f-5fc9j 
{"status":"ok","version":"UNKNOWN"}
GraphQL service healthy!

> @ serve /apollo
> node dist/index.js

Building schema...
Error: Could not obtain introspection result, received: {"path":"$","error":"resource does not exist","code":"not-found"}
    at getSchemaFromIntrospection (/apollo/node_modules/@graphql-tools/wrap/index.cjs.js:1370:15)
    at introspectSchema (/apollo/node_modules/@graphql-tools/wrap/index.cjs.js:1379:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Trying again in 3 seconds...
Building schema...
Error: Could not obtain introspection result, received: {"path":"$","error":"resource does not exist","code":"not-found"}
    at getSchemaFromIntrospection (/apollo/node_modules/@graphql-tools/wrap/index.cjs.js:1370:15)
    at introspectSchema (/apollo/node_modules/@graphql-tools/wrap/index.cjs.js:1379:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Trying again in 3 seconds...
```

## Changes

Adds experimental helm chart.

## Importance

Allows deployment of prefect server to kubernetes.

## Checklist

This PR:

- [ ] adds new tests (if appropriate)
- [ ] adds a change file in the `changes/` directory (if appropriate)

### CODE REVIEW

**Comment by jlowin** [comment]
@shaunc sorry for the delay -- I'm curious if the failure is on the apollo side or the graphql side. Essentially, Apollo is making the introspection request to learn the graphql server's schema -- do you see that request arriving in the graphql pod's logs? Or is the apollo pod unable to reach the graphql pod at all?

**Comment by shaunc** [comment]
Looks like its not getting through:
graphql logs are filled with:
```
INFO:     192.168.192.0:34644 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34680 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34696 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34740 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34754 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34792 - "GET /health HTTP/1.1" 200 OK
INFO:     192.168.192.0:34814 - "GET /health HTTP/1.1" 200 OK
```
[But ... hmm... is the log level ok ...?]
... I will try to override the entrypoint so I can try to connect from the Apollo pod itself

**Comment by shaunc** [comment]
Ok -- replacing the pod command with `sleep 999999` and `exec`ing into the shell, I have:
```
$ env | grep GRAPHQL
PREFECT_SERVER_GRAPHQL_PORT_4201_TCP=tcp://172.20.247.123:4201
PREFECT_SERVER_GRAPHQL_PORT=tcp://172.20.247.123:4201
PREFECT_SERVER_GRAPHQL_SERVICE_PORT_HTTP=4201
PREFECT_SERVER_GRAPHQL_PORT_4201_TCP_PORT=4201
PREFECT_SERVER__SERVICES__GRAPHQL__PORT=4201
PREFECT_SERVER_GRAPHQL_SERVICE_PORT=4201
PREFECT_SERVER_GRAPHQL_PORT_4201_TCP_ADDR=172.20.247.123
PREFECT_SERVER_GRAPHQL_PORT_4201_TCP_PROTO=tcp
GRAPHQL_SERVICE_PORT=4201
PREFECT_SERVER_GRAPHQL_SERVICE_HOST=172.20.247.123
GRAPHQL_SERVICE_HOST=http://prefect-server-graphql.prefect-server.svc.cluster.local
```
The following work:
```
$ curl http://prefect-server-graphql.prefect-server.svc.cluster.local:4201/health
{"status":"ok","version":"UNKNOWN"}
$ curl http://172.20.247.123:4201/health
{"status":"ok","version":"UNKNOWN"}
```
The "TCP" env look odd. ... not in the deployment env -- which is (I autogenerate for all pods, so contains settings "for whole application"):
```
        - name: PREFECT_SERVER__HASURA__HOST
          value: prefect-server-hasura.prefect-server.svc.cluster.local
        - name: PREFECT_SERVER__HASURA__PORT
          value: "3000"
        - name: PREFECT_SERVER__SERVICES__APOLLO__PORT
          value: "4200"
        - name: PREFECT_SERVER__SERVICES__GRAPHQL__PORT
          value: "4201"
        - name: PREFECT_SERVER__DEBUG
          value: "false"
        - name: PREFECT_SERVER__ENV
          value: local
        - name: PREFECT_SERVER__HASURA__EXECUTE_RETRY_SECONDS
          value: "10"
        - name: PREFECT_SERVER__LOGGING__FORMAT
          value: '[%(asctime)s] %(levelname)s - %(name)s | %(message)s'
        - name: PREFECT_SERVER__LOGGING__LEVEL
          value: DEBUG
        - name: PREFECT_SERVER__QUEUED_RUNS_RETURNED_LIMIT
          value: "25"
        - name: PREFECT_SERVER__SERVICES__LAZARUS__RESURRECTION_ATTEMPT_LIMIT
          value: "3"
        - name: PREFECT_SERVER__SERVICES__SCHEDULER__SCHEDULER_LOOP_SECONDS
          value: "300"
        - name: PREFECT_SERVER__TELEMETRY__ENABLED
          value: "true"
```
Don't know what this is? Should it be something else for kubernetes?
```
        - name: PREFECT_SERVER__ENV
          value: local
```

**Comment by shaunc** [comment]
Also I have:
```
env | grep PREFECT_API
PREFECT_API_URL=http://prefect-server-graphql.prefect-server.svc.cluster.local:4201/graphql/
PREFECT_API_HEALTH_URL=http://prefect-server-graphql.prefect-server.svc.cluster.local:4201/health
```
And 
```
curl $PREFECT_API_URL
```
works, and results in log entry at graphql:
```
INFO:     192.168.192.11:37670 - "GET /graphql/ HTTP/1.1" 200 OK
```

**Comment by joshmeek** [comment]
Hi @shaunc I fixed the issue you were encountering however it won't let me push to your branch.

image

```
# apollo logs
{"status":"ok","version":"UNKNOWN"}
GraphQL service healthy!

> @ serve /apollo
> node dist/index.js

Building schema...
Building schema complete!
Server ready at http://0.0.0.0:4200 🚀 (version: UNKNOWN)
Sending telemetry to Prefect Technologies, Inc.: {"source":"prefect_server","type":"startup","payload":{"id":"de3310a9-1826-48d1-b93b-813714d8b413","prefect_server_version":"UNKNOWN","api_version":"0.2.0"}}
```

The hasura helpers `api-url` has it set to `v1alpha` and it should be `v1alpha1`

**Comment by shaunc** [comment]
Josh -- thanks so much. I added you as able to push to the source repo -- factfiber/server

File: helm/prefect-server/templates/hasura/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+{{- $name := include "hasura.fullname" . -}}
+{{- $ns := include "global.namespace" . }}
+{{- $suffix := .Values.global.fqdnSuffix }}
+{{- printf "%s.%s.%s" $name $ns $suffix -}}
+{{- end -}}
+
+{{- define "hasura.api-url" -}}
+{{- $host := include "hasura.fqdn" . -}}
+{{- $port := .Values.global.hasura.port | toString -}}
+{{ printf "http://%s:%s/v1alpha/graphql" $host $port }}
```
Comment (joshmeek) [inline helm/prefect-server/templates/hasura/_helpers.tpl:19]:
```suggestion
{{ printf "http://%s:%s/v1alpha1/graphql" $host $port }}
```

Comment (joshmeek) [reply helm/prefect-server/templates/hasura/_helpers.tpl:19]: Still wasn't letting my push, nbd I can figure it out later. Here's the change ^

**Comment by shaunc** [comment]
Cool! Thanks!

**Comment by yolibernal** [comment]
We are currently using the Helm charts and they work great so far. The only two issues we encountered were

1) Setting the resources in `values.yaml` failed with `mapping values are not allowed in this context`. To resolve we changed `{{- with .Values.hasura.resources -}}` to `{{- with .Values.hasura.resources }}` (removed trailing slash, similar to [here](https://github.com/helm/helm/issues/6251#issuecomment-528064834), don't know if it is the same error)
2) The deployments had to be named `prefect-server` probably due to [here](https://github.com/PrefectHQ/server/pull/57/files#diff-e79c366c135a6598b07f5e87a5d81a66b618f3d30ce29c33891aabdafcfca625R89) which already is marked as a TODO

File: helm/prefect-server/templates/apollo/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  tls:
+  {{- range .Values.apollo.ingress.tls }}
+    - hosts:
+      {{- range .hosts }}
+        - {{ . | quote }}
+      {{- end }}
+      secretName: {{ .secretName }}
+  {{- end }}
+{{- end }}
+  {{- if .Values.apollo.ingress.hosts  -}}
```
Comment (joelluijmes) [inline helm/prefect-server/templates/apollo/ingress.yaml:28]:
```suggestion
  {{- if .Values.apollo.ingress.hosts }}
```

The trimming - breaks the template

File: helm/prefect-server/templates/ui/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  tls:
+  {{- range .Values.ui.ingress.tls }}
+    - hosts:
+      {{- range .hosts }}
+        - {{ . | quote }}
+      {{- end }}
+      secretName: {{ .secretName }}
+  {{- end }}
+{{- end }}
+  {{- if .Values.ui.ingress.hosts  -}}
```
Comment (joelluijmes) [inline helm/prefect-server/templates/ui/ingress.yaml:28]:
```suggestion
  {{- if .Values.ui.ingress.hosts }}
```

The trimming - breaks template

**Review by joelluijmes — COMMENTED** [review]
I was looking for a helm chart for Prefect, and I found this PR 💯

Don't got it quite working yet personally, the generated ingress is not correct (or I'm providing values incorrectly, my paths aren't generated as I expected).. Anyhow also found a bug due trimming `-}}`.

Anyhow, I really liked to see this merged 👍

**Comment by joelluijmes** [comment]
> Don't got it quite working yet personally, the generated ingress is not correct (or I'm providing values incorrectly, my paths aren't generated as I expected).

Okay, fixed the generation of the ingress yaml. Here is the full patch in case you want to include it 👍 

```patch
diff --git a/helm/prefect-server/templates/ui/ingress.yaml b/helm/prefect-server/templates/ui/ingress.yaml
index 1e1a3ca..f34ae4f 100644
--- a/helm/prefect-server/templates/ui/ingress.yaml
+++ b/helm/prefect-server/templates/ui/ingress.yaml
@@ -1,4 +1,5 @@
 {{- if .Values.ui.ingress.enabled -}}
+{{- $fullname := include "ui.fullname" . -}}
 {{- $svcPort := .Values.global.ui.port -}}
 {{- if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
 apiVersion: networking.k8s.io/v1beta1
@@ -25,16 +26,17 @@ spec:
       secretName: {{ .secretName }}
   {{- end }}
 {{- end }}
-  {{- if .Values.ui.ingress.hosts  -}}
+  {{- if .Values.ui.ingress.hosts }}
   rules:
     {{- range .Values.ui.ingress.hosts }}
     - host: {{ .host | quote }}
+      random: {{ .random }}
       http:
         paths:
         {{- range .paths }}
           - path: {{ . }}
             backend:
-              serviceName: {{ include "ui.fullname" . }}
+              serviceName: {{ $fullname }}
               servicePort: {{ $svcPort }}
         {{- end }}
     {{- end }}
```

**Comment by madkinsz** [comment]
Closing in favor of #123, thanks everyone for your work here!

**Pull request closed without merging**